### ISSUE TITLE
Mod suggestion: Better Advancements

### ISSUE THREAD

**Issue by EliteMasterEric** [martmists-gh/fabric-mod-wishlist#54]
### Description

Better Advancements is a concept which expands the built-in functionality of Minecraft's basic achievement system. The goal here is to transform them into a fully featured questing system, in the vein of Better Questing, FTB Quests, and Hardcore Questing Mode. These mods provide an important functionality to modpack creators, allowing them to define goals and rewards for players based on certain triggers, which has been a major hinderance in widespread adoption of Fabric. Also, in my personal opinion, these Forge-based questing mods have ugly user interfaces that clash with default Minecraft and take up too much of the viewport.

The default advancements, as part of the datapack system, are already surprisingly flexible; they provide functionality for names, descriptions, item-based icons, a GUI with positioned icons, criteria, prerequisites, hidden advancements, multiple conditions, a tabbed category view, custom backgrounds, and rewards, and any user can access them easily through the pause menu without crafting a book. Currently, the reward system has functionality to run a set of commands (through .mcfunction files), provide randomized loot via loot tables, provide an amount of XP, or even unlock recipes, which has a lot of potential when `/gamerule doLimitedCrafting` is enabled.

However, in several areas, advancements are not feasible to be used by mod authors to create a mod pack centered around them. My proposal includes new functionality to make advancements fully-featured enough to use for questing. The mod would provide the following functionalities:

* Currently, the only information on an advancement is a brief description. Add functionality such that, when an advancement is clicked, a generated description of the triggers required to complete the advancement appears, along with an extended description if available.

* Currently, advancements only have one set of simultanous requirements. Allow for certain advancements to be completed in multiple stages, such as obtaining one item, then obtaining another item.

* Currently, advancement rewards are granted instantly upon being completed, rather than when the user acknowledges and claims them. This can be problematic if the user's inventory is full. This mod would (by default) require users to access the Advancement menu and click the Claim button to gain the rewards.

* Currently, advancements cannot require the user to spend certain items. As part of the Claim functionality, allow a new type of advancement which requires the user to submit one or more items in exchange for completion and rewards.

* Currently, advancements cannot be repeated. Adding repeated reward claiming, tied in with advancement costs as described above, allows for certain advancements to repeatedly provide certain resources in exchange for another resource, as is often done with certain other questing modpacks like Project Ozone 3.

* Currently, it is difficult to edit advancements, and this cannot be done in-game. Add an in-game GUI-based editor, enabling modpack authors to create advancements without leaving the game, editing raw JSON, then reloading to test. This may merit being part of a separate mod for the sake of modularity and compatibility.

* Currently, advancements cannot currently be completed as a group of players. Include a party management system allowing players to claim advancements together. This may merit being part of a separate mod as well.

* Add additional types of advancement triggers as needed to allow for more complex advancements. This may merit being part of a separate mod as well.

TL;dr A questing mod based on the existing advancement system.

### Details

- Does a Forge version of this mod exist? No, but several other questing mods exist, implementing their own interfaces.
- Does a Fabric version of this mod exist for an older MC version? No

**Comment by ehx-v1** [comment]
> Currently, advancements cannot be repeated.

Oh yes, they can. Simply include an advancement-revoking function in the reward. You can have multiple rewards per advancement, so no potential is wasted.

> Currently, advancements cannot currently be completed as a group of players.

Possible as well, and the solution are, once again, reward functions. Have the reward function grant the advancement to all players. To be fair though, that only works for one-off advancements and not for Adventuring Time styled long-running questlines (although that could be solved with an `advancement_cleared` trigger and no-display helper advancements).

Other than that, full support from me, especially for the "more triggers" part. In fact, I feel like `scoreboard_changed`, with full value check support as described [here](https://www.minecraftforum.net/forums/minecraft-java-edition/suggestions/2870843-advancement-trigger-scoreboard_changed), and `advancement_cleared` allowing to refer to specific other advancements, would (in combination with the rest of the datapack system) be extremely powerful and might even cover the entire request except for the in-game UI (although it would be insanely complicated to do some things, so some QoL shortcut features would still be neat). Oh, and if the mod were to support 1.15, pre-porting `used_item_on_block` is obviously mandatory.
